src/shared/codelens/pythonCodeLensProvider.ts

```diff
@@ -59,75 +70,210 @@ const getLambdaHandlerCandidates = async ({ uri }: { uri: vscode.Uri }): Promise
         })
 }
 
-export function initialize({
+// Add create debugging manifest/requirements.txt containing ptvsd
+const makePythonDebugManifest = async (params: {
+    samProjectCodeRoot: string,
+    outputDir: string
+}): Promise<string | undefined> => {
+    let manifestText = ''
+    const manfestPath = path.join(params.samProjectCodeRoot, 'requirements.txt')
+    if (fileExists(manfestPath)) {
+        manifestText = await readFileAsString(manfestPath)
+    }
+    getLogger().debug(`pythonCodeLensProvider.makePythonDebugManifest params: ${JSON.stringify(params, undefined, 2)}`)
+    if (manifestText.indexOf('ptvsd') < 0 ) { // TODO: Make this logic more robust
+        manifestText += '\nptvsd'
+        const debugManifestPath = path.join(params.outputDir, 'debug-requirements.txt')
+        await writeFile(debugManifestPath, manifestText)
+
+        return debugManifestPath
+    }
+    // else we don't need to override the manifest. nothing to return
+}
+
+// tslint:disable:no-trailing-whitespace
+const makeLambdaDebugFile = async (params: {
+    handlerName: string,
+    debugPort: number,
+    outputDir: string
+}): Promise<{outFilePath: string, debugHandlerName: string}> => {
+    if (!params.outputDir) {
+        throw new Error('Must specify outputDir')
+    }
+    const logger = getLogger()
+
+    const [handlerFilePrefix, handlerFunctionName] = params.handlerName.split('.')
+    const debugHandlerFileName = `${handlerFilePrefix}_debug`
+    const debugHandlerFunctionName = 'lambda_handler'
+    try {
+        logger.info('pythonCodeLensProvider.makeLambdaDebugFile params:', JSON.stringify(params, undefined, 2))
+        const template = `
+import ptvsd
+from ${handlerFilePrefix} import ${handlerFunctionName} as _handler
+
+
+def ${debugHandlerFunctionName}(event, context):
+    print('waiting for debugger to attach...')
+    ptvsd.enable_attach(address=('0.0.0.0', ${params.debugPort}), redirect_output=True)
+    ptvsd.wait_for_attach()
+    print('debugger attached')
+    return _handler(event, context)
+
+`
+
+        const outFilePath = path.join(params.outputDir, `${debugHandlerFileName}.py`)
+        logger.info('pythonCodeLensProvider.makeLambdaDebugFile outFilePath:', outFilePath)
+        await writeFile(outFilePath, template)
```

Comment:
What happens to this file after debugging? Do we clean it up?

Reply:
Updated to use `ExtensionDisposableFiles` to delete it. Depending upon how messy Python is, it may be helpful to leave it there and inform the user they can use it for debugging purposes... but hopefully we can do better than that.
✔️

Reply:
Actually the manifest is in `makeBuildDir` but `lambdaDebugFile` is in the SAM code dir because I haven't yet been able to get it to wok elsewhere. I do add it to `ExtensionDisposableFiles` though.

Reply:
That won't clean it up until the user closes VS Code, right? Can we clean it up after the debug session?

Reply:
If we wanted to but I don't see that as a blocker.

Reply:
Updated to immediately delete `lambdaDebugFile` after Lambda invokation.
✔